### DISCUSSION TITLE
@breakpoint annotation

### DISCUSSION THREAD

**Discussion by bmaschio** [jolie/jolie#576] [Ideas (Features, Fixes, etc.)]
Hi everyone

I was talking the other day with @klag  about the need of having a break-point  mechanism to debug  i suggested the use of keyword  maybe with some kind of annotation such as 

@breakpoint 

this will be seen only when  jolie --trace name.ol
We could couple it with Atom and in some kind of way pipe it back to the editor
What you think  ?

**Comment by fmontesi** [comment]
I'm not fully convinced that this should be in the language. Isn't this usually just dealt with by the debugging tool? For example, by telling it that you want a breakpoint at line 50 or something like that.

**Comment by bmaschio** [comment]
@fmontesi how would you pass a series of break point to the parser?  my idea was something like this 
https://atom.io/packages/php-debug not sure how to go about it

**Comment by fmontesi** [comment]
I would just edit all process implementations, e.g., `SequentialProcess`, such that they ask the interpreter permission to proceed (and possibly wait because of it) before taking the next step or iteration. Then, the interpreter does a simple check of whether a Debugger service has been attached. This should be a service, just like we do with monitors right now. (Actually, maybe we should just extend monitors such that they can make the interpreter wait?) Sending something to the debugger should be a request-response, so that the debugger may decide that the interpreter should wait.

Does that make sense?

**Comment by bmaschio** [comment]
@fmontesi I like it the idea of expanding the monitor because if I am not mistaken  we can read the thread variable value. How would you handle multi threading and call chains  op1@micro1=> op2@micro2 => op3@micro3 ?

**Comment by fmontesi** [comment]
Multi-threading: That's a challenge, we should probably get some
inspiration from the debugger in netbeans. But basically, we'll have to
deal with the fact that multiple processes would be running at the same
time.

Call chains: Not sure what you mean. A debugger is attached to a single
service. If you want to debug another service, then it'll have its own
debugger attached. I'd think on how to connect multiple debuggers later on.
